src/rgw/rgw_kafka.cc

```diff
@@ -377,11 +381,11 @@ class Manager {
       topic = rd_kafka_topic_new(conn->producer, message->topic.c_str(), nullptr);
       if (!topic) {
         const auto err = rd_kafka_last_error();
-        ldout(conn->cct, 1) << "Kafka publish: failed to create topic: " << message->topic << " error: " << status_to_string(err) << dendl;
+        ldout(conn->cct, 1) << "Kafka publish: failed to create topic: " << message->topic << " error: " 
```

Comment:
this whole logic of calling `rd_kafka_topic_new` if topic does not exist doesn't make sense.
what could be the scenario where this could happen.
`publish_internal` will always be called after the connection is created ?

Reply:
from kafka perspective "connection" and "topic" are not the same. one connection may serve multiple topics.
if we try to publish to a topic and it does not exist, publishing would fail.
so, to be on the safe side, we create the topic (the kafka topic) first.

Reply:
sorry i think i misinterpreted this, 
so kafka needs to call `rd_kafka_topic_new` before calling `rd_kafka_produce`.
but what is confusing is why is this a vector ?
for a given kafka topic, what are the scenario where we will need multiple `rd_kafka_topic_new` ?
this vector list i weird, the  vector is holding `rd_kafka_topic_t` --> `std::vector<rd_kafka_topic_t*> topics`
and while searching `auto topic_it = std::find(conn->topics.begin(), conn->topics.end(), message->topic);` you are searching for a `string` inside the vector ?

so basically i am asking is
-> why do we need vector ? a given topic will have one on one mapping to `rd_kafka_topic_t`, so while we connection caching(pooling) we create the connection and then we need to call `rd_kafka_topic_new` only once ?

Reply:
an RGW topic contains 2 main things: the endpoint and the topic name.
the kafka manager holds a list of connection (=endpoints) each has a vector of kafka topics.
lets assume we have the following scenarion:
RGW topic1: kafka://host1:9092, mytopic1
RGW topic2: kafka://host1:9092, mytopic2

so, in the kafka manager, we will have one connection (to host1:9092) with a vector of 2 topics: topic1, topic2
when we send the notifications, once we created the 2 topics on the broker, we will see that they are in the vector, and we won't create them again.
note, that when the RGW topics are created, the kafka manager has no information about that. only when it sends the notifications it opens the connection and creates the topics

Reply:
got it, its just the name of the topic that is passed during the `CreateTopic`
but again why maintain the vector ? coz they both are pointing to same connection? can't both topics use the same connection ?

Reply:
inside the kafka manager we do not masintain a list of RGW topics.
we maintain a list of kafka endpoints, and each endpoint is holding a vector of topics.
this means that if we have multipel RGW topics that share a connection, they will become a single connection with a vector of topic names in the kafka manager

Reply:
Yep I got that logic, Kafka manager topic can hold multiple rgw topics , what I was asking is why can't multiple rgw topics map to single Kafka connection ?
Is it necessary to maintain this vector ?

Reply:
The kafka manager does not know anything about rgw topics.
It hold a list of connection (=brokers). Each holding a list of kafka topics pn that broker.

Reply:
> The kafka manager does not know anything about rgw topics. It hold a list of connection (=brokers). Each holding a list of kafka topics pn that broker.

@yuvalif we are repeating the same thing,
kafka manager maintains a map based on endpoint/broker.
the 2 rgw topics point to same endpoint/broker, so both rgw topics can technically use the same connection if we do not create the vector? is there any reason we map 2 rgw topics to single kafka manager connection? can't we use same